### PR TITLE
[SPARK-37066][SQL] Improve error message to show file path when failed to read next file

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### What changes were proposed in this pull request?
When use ORC vectorized reader,  if `OrcColumnarBatchReader` throw `ArrayIndexOutofBoundsException`, it only show error message of index, won't show which file is being reading
```

21/10/19 18:12:58 ERROR Executor: Exception in task 34.1 in stage 14.0 (TID 257)
java.lang.ArrayIndexOutOfBoundsException: 1024
	at org.apache.orc.impl.TreeReaderFactory$TreeReader.nextVector(TreeReaderFactory.java:292)
	at org.apache.orc.impl.TreeReaderFactory$StringDictionaryTreeReader.nextVector(TreeReaderFactory.java:1820)
	at org.apache.orc.impl.TreeReaderFactory$StringTreeReader.nextVector(TreeReaderFactory.java:1517)
	at org.apache.orc.impl.ConvertTreeReaderFactory$DateFromStringGroupTreeReader.nextVector(ConvertTreeReaderFactory.java:1802)
	at org.apache.orc.impl.TreeReaderFactory$StructTreeReader.nextBatch(TreeReaderFactory.java:2059)
	at org.apache.orc.impl.RecordReaderImpl.nextBatch(RecordReaderImpl.java:1324)
	at org.apache.spark.sql.execution.datasources.orc.OrcColumnarBatchReader.nextBatch(OrcColumnarBatchReader.java:196)
	at org.apache.spark.sql.execution.datasources.orc.OrcColumnarBatchReader.nextKeyValue(OrcColumnarBatchReader.java:99)
	at org.apache.spark.sql.execution.datasources.RecordReaderIterator.hasNext(RecordReaderIterator.scala:37)
	at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1.hasNext(FileScanRDD.scala:93)
	at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1.nextIterator(FileScanRDD.scala:173)
	at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1.hasNext(FileScanRDD.scala:93)
	at org.apache.spark.sql.execution.FileSourceScanExec$$anon$1.hasNext(DataSourceScanExec.scala:503)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage1.columnartorow_nextBatch_0$(Unknown Source)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage1.processNext(Unknown Source)
	at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
	at org.apache.spark.sql.execution.WholeStageCodegenExec$$anon$1.hasNext(WholeStageCodegenExec.scala:755)
	at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:458)
	at org.apache.spark.shuffle.sort.BypassMergeSortShuffleWriter.write(BypassMergeSortShuffleWriter.java:132)
	at org.apache.spark.shuffle.ShuffleWriteProcessor.write(ShuffleWriteProcessor.scala:59)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:99)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:52)
	at org.apache.spark.scheduler.Task.run(Task.scala:131)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:497)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1439)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:500)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

It caused difficult for SRE/Admin to find out which file have problem. 

In this pr, we add a catch in `FileScanRDD` & `FilePartitionReader` to log an error message to show the file path when failed read data.

### Why are the changes needed?
Improve error message for orc vectorized reader

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?

Existed UT

After this patch, error message in our PROD
```
21/10/22 16:01:39 WARN TaskSetManager: Lost task 312.0 in stage 1.0 (TID 55) (ip-10-130-15-162.idata-server.shopee.io executor 30): org.apache.spark.sql.execution.QueryExecutionException: Encountered error while reading Parquet file hdfs://path/to/file/part-00001-2d12874d-36b5-4091-9202-3dc0c81c2d9b-c000.snappy.orc. Details:
	at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1.nextIterator(FileScanRDD.scala:193)
	at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1.hasNext(FileScanRDD.scala:93)
	at org.apache.spark.sql.execution.FileSourceScanExec$$anon$1.hasNext(DataSourceScanExec.scala:503)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage1.columnartorow_nextBatch_0$(Unknown Source)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage1.agg_doAggregateWithoutKey_0$(Unknown Source)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage1.processNext(Unknown Source)
	at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
	at org.apache.spark.sql.execution.WholeStageCodegenExec$$anon$1.hasNext(WholeStageCodegenExec.scala:755)
	at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:458)
	at org.apache.spark.shuffle.sort.BypassMergeSortShuffleWriter.write(BypassMergeSortShuffleWriter.java:132)
	at org.apache.spark.shuffle.ShuffleWriteProcessor.write(ShuffleWriteProcessor.scala:59)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:99)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:52)
	at org.apache.spark.scheduler.Task.run(Task.scala:131)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:497)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1439)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:500)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.ArrayIndexOutOfBoundsException: 1024
	at org.apache.orc.impl.TreeReaderFactory$TreeReader.nextVector(TreeReaderFactory.java:292)
	at org.apache.orc.impl.TreeReaderFactory$StringDictionaryTreeReader.nextVector(TreeReaderFactory.java:1820)
	at org.apache.orc.impl.TreeReaderFactory$StringTreeReader.nextVector(TreeReaderFactory.java:1517)
	at org.apache.orc.impl.ConvertTreeReaderFactory$DateFromStringGroupTreeReader.nextVector(ConvertTreeReaderFactory.java:1802)
	at org.apache.orc.impl.TreeReaderFactory$StructTreeReader.nextBatch(TreeReaderFactory.java:2059)
	at org.apache.orc.impl.RecordReaderImpl.nextBatch(RecordReaderImpl.java:1324)
	at org.apache.spark.sql.execution.datasources.orc.OrcColumnarBatchReader.nextBatch(OrcColumnarBatchReader.java:196)
	at org.apache.spark.sql.execution.datasources.orc.OrcColumnarBatchReader.nextKeyValue(OrcColumnarBatchReader.java:99)
	at org.apache.spark.sql.execution.datasources.RecordReaderIterator.hasNext(RecordReaderIterator.scala:37)
	at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1.hasNext(FileScanRDD.scala:93)
	at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1.nextIterator(FileScanRDD.scala:173)
	... 19 more
```

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
How was this patch tested?

**Comment by dongjoon-hyun** [comment]
Thank you for pinging me, @AngersZhuuuu .

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
It would be great if we can have this exception handling in `FileScanRDD.scala`. I believe we can use `currentFile.filePath` instead of introducing another string variable to ORC/Parquet file format seperately. Could you try that way, @AngersZhuuuu ?

```scala
          try {
            hasNext
          } catch {
            case e: SchemaColumnConvertNotSupportedException =>
              throw QueryExecutionErrors.unsupportedSchemaColumnConvertError(
                currentFile.filePath, e.getColumn, e.getLogicalType, e.getPhysicalType, e)
            case e: ParquetDecodingException =>
              if (e.getCause.isInstanceOf[SparkUpgradeException]) {
                throw e.getCause
              } else if (e.getMessage.contains("Can not read value at")) {
                throw QueryExecutionErrors.cannotReadParquetFilesError(e)
              }
              throw e
          }
```

**Comment by dongjoon-hyun** [comment]
Thank you for generalizing and updating, @AngersZhuuuu . Could you check the UT failures? It looks like wrapping exceptions requires us revise the test case asserts.
```
[info] - Logical type: Decimal with too large precision *** FAILED *** (65 milliseconds)
[info]   "Encounter error while reading parquet file file:///home/runner/work/spark/spark/target/tmp/spark-234e31e3-7e34-47a5-90bf-5f716c0e1e6b.avro.  Details: " did not contain "Unscaled value too large for precision" (AvroLogicalTypeSuite.scala:438)
```

**Comment by AngersZhuuuu** [comment]
> Thank you for generalizing and updating, @AngersZhuuuu . Could you check the UT failures? It looks like wrapping exceptions requires us revise the test case asserts.
> 
> ```
> [info] - Logical type: Decimal with too large precision *** FAILED *** (65 milliseconds)
> [info]   "Encounter error while reading parquet file file:///home/runner/work/spark/spark/target/tmp/spark-234e31e3-7e34-47a5-90bf-5f716c0e1e6b.avro.  Details: " did not contain "Unscaled value too large for precision" (AvroLogicalTypeSuite.scala:438)
> ```

Done

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FilePartitionReader.scala [hunk 1]
```diff
@@ -66,17 +64,16 @@ class FilePartitionReader[T](readers: Iterator[PartitionedFileReader[T]])
...
-          throw e.getCause
-        } else if (e.getMessage.contains("Can not read value at")) {
-          throw QueryExecutionErrors.cannotReadParquetFilesError(e)
-        }
-        throw e
       case e @ (_: RuntimeException | _: IOException) if ignoreCorruptFiles =>
         logWarning(
           s"Skipped the rest of the content in the corrupted file: $currentReader", e)
         false
+      case e: Exception =>
```
Comment (srowen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FilePartitionReader.scala:71]: No big deal, but these could be:
```
e.getCause match {
  case sue: SparkUpgradeException => throw sue
  case _ => throw ...
}
```

**Review by srowen — COMMENTED** [review]
Seems OK

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FilePartitionReader.scala (hunk 1 above)
Comment (AngersZhuuuu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FilePartitionReader.scala:71]: > No big deal, but these could be:
> 
> ```
> e.getCause match {
>   case sue: SparkUpgradeException => throw sue
>   case _ => throw ...
> }
> ```

Good Suggestion, updated

**Comment by AngersZhuuuu** [comment]
ping @cloud-fan

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileScanRDD.scala [hunk 2]
```diff
@@ -170,13 +168,12 @@ class FileScanRDD(
             case e: SchemaColumnConvertNotSupportedException =>
               throw QueryExecutionErrors.unsupportedSchemaColumnConvertError(
                 currentFile.filePath, e.getColumn, e.getLogicalType, e.getPhysicalType, e)
-            case e: ParquetDecodingException =>
-              if (e.getCause.isInstanceOf[SparkUpgradeException]) {
-                throw e.getCause
-              } else if (e.getMessage.contains("Can not read value at")) {
-                throw QueryExecutionErrors.cannotReadParquetFilesError(e)
+            case e: Exception =>
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileScanRDD.scala:171]: I think `NonFatal(e)` is more commonly used?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FilePartitionReader.scala [hunk 3]
```diff
@@ -66,17 +64,16 @@ class FilePartitionReader[T](readers: Iterator[PartitionedFileReader[T]])
...
+        e.getCause match {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FilePartitionReader.scala:72]: ditto

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileScanRDD.scala (hunk 2 above)
Comment (AngersZhuuuu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileScanRDD.scala]: Done

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRebaseDatetimeSuite.scala [hunk 4]
```diff
@@ -164,7 +165,13 @@ abstract class ParquetRebaseDatetimeSuite
     def successInRead(path: String): Unit = spark.read.parquet(path).collect()
     def failInRead(path: String): Unit = {
       val e = intercept[SparkException](spark.read.parquet(path).collect())
-      assert(e.getCause.isInstanceOf[SparkUpgradeException])
+      // If SparkUpgradeException is thrown while reading file,
+      // it will be wrapped by QueryExecutionException.
+      if (e.getCause.isInstanceOf[QueryExecutionException]) {
```
Comment (srowen) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRebaseDatetimeSuite.scala:170]: Do we want the same match pattern everywhere else, like here?

Comment (srowen) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRebaseDatetimeSuite.scala:170]:
```
e.getCause match {
  case QueryExecutionException => assert(...)
  case SparkUpgradeException => // OK
  case _ => fail(...)
}
```
Yeah maybe not cleaner, it's OK

Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRebaseDatetimeSuite.scala:170]: BTW, where do we wrap `SparkUpgradeException`? In `FileScanRDD` we intentionally avoid wrapping `SparkUpgradeException`.

Comment (AngersZhuuuu) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRebaseDatetimeSuite.scala:170]: > BTW, where do we wrap `SparkUpgradeException`? In `FileScanRDD` we intentionally avoid wrapping `SparkUpgradeException`.

Seems origin code
```
case e: ParquetDecodingException =>
        if (e.getCause.isInstanceOf[SparkUpgradeException]) {
          throw e.getCause
        } else if (e.getMessage.contains("Can not read value at")) {
          throw QueryExecutionErrors.cannotReadParquetFilesError(e)
        }
        throw e
```
can't catch SparkUpgradeException, SparkUpgradeException is directly thrown, not wrapped by ParquetDecodingException

Comment (AngersZhuuuu) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRebaseDatetimeSuite.scala]: > BTW, where do we wrap `SparkUpgradeException`? In `FileScanRDD` we intentionally avoid wrapping `SparkUpgradeException`.

Handle this,

Comment (AngersZhuuuu) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRebaseDatetimeSuite.scala]: > ```
> e.getCause match {
>   case QueryExecutionException => assert(...)
>   case SparkUpgradeException => // OK
>   case _ => fail(...)
> }
> ```
> 
> Yeah maybe not cleaner, it's OK

Won't need this change now.

**Comment by srowen** [comment]
Merged to master

**Review by dongjoon-hyun — COMMENTED** [review]
+1, late LGTM. Thank you all.

**Pull request closed without merging**